generator/booking_dataset.cpp

```diff
@@ -0,0 +1,242 @@
+  for (size_t i = 0; token && i < rec.size(); ++i, ++token)
+    rec[i] = *token;
+
+  strings::to_uint(rec[Index(Fields::Id)], id);
+  strings::to_double(rec[Index(Fields::Latitude)], lat);
+  strings::to_double(rec[Index(Fields::Longtitude)], lon);
+  
+  name = rec[Index(Fields::Name)];
+  address = rec[Index(Fields::Address)];
+
+  strings::to_uint(rec[Index(Fields::Stars)], stars);
+  strings::to_uint(rec[Index(Fields::PriceCategory)], priceCategory);
+  strings::to_double(rec[Index(Fields::RatingBooking)], ratingBooking);
+  strings::to_double(rec[Index(Fields::RatingUsers)], ratingUser);
+
+  descUrl = rec[Index(Fields::DescUrl)];
+
+  strings::to_uint(rec[Index(Fields::Type)], type);
+}
+
+ostream & operator<<(ostream & s, BookingDataset::Hotel const & h)
+{
+  return s << "Name: " << h.name << " lat: " << h.lat << " lon: " << h.lon;
+}
+
+void BookingDataset::LoadHotels(string const & path)
+{
+  m_hotels.clear();
+
+  if (path.empty())
+    return;
+
+  ifstream src(path);
+  for (string line; getline(src, line);)
+    m_hotels.emplace_back(line);
+}
+
+BookingDataset::BookingDataset(string const & dataPath)
+{
+  LoadHotels(dataPath);
+
+  size_t counter = 0;
+  for (auto const & hotel : m_hotels)
+  {
+    TBox b(TPoint(hotel.lat, hotel.lon), TPoint(hotel.lat, hotel.lon));
+    m_rtree.insert(std::make_pair(b, counter));
+    ++counter;
+  }
+}
+
+bool CheckForValues(string const & value)
+{
+  for (char const * val :
+       {"hotel", "apartment", "camp_site", "chalet", "guest_house", "hostel", "motel", "resort"})
+  {
+    if (value == val)
+      return true;
+  }
+  return false;
+}
+
+bool BookingDataset::MatchWithBooking(OsmElement const & e) const
+{
+  string name;
+  for (auto const & tag : e.Tags())
+  {
+    if (tag.key == "name")
+    {
+      name = tag.value;
+      break;
+    }
+  }
+
+  if (name.empty())
+    return false;
+
+  // Find 3 nearest values to a point.
+  vector<TValue> result;
+  for_each(boost::geometry::index::qbegin(m_rtree,
+                                          boost::geometry::index::nearest(TPoint(e.lat, e.lon), 3)),
+           boost::geometry::index::qend(m_rtree), [&](TValue const & v)
+           {
+             auto const & hotel = m_hotels[v.second];
+             double dist = ms::DistanceOnEarth(e.lat, e.lon, hotel.lat, hotel.lon);
+             if (dist > 150 /* max distance in meters */)
+               return;
+
+             result.emplace_back(v);
+           });
+
+  if (result.empty())
+    return false;
+
+  // Match name.
+  vector<strings::UniString> osmTokens;
+  NormalizeAndTokenizeString(name, osmTokens, search::Delimiters());
+
+  //  cout << "\n------------- " << name << endl;
+
+  bool matched = false;
+  for (auto const & e : result)
+  {
+    vector<strings::UniString> bookingTokens;
+    NormalizeAndTokenizeString(m_hotels[e.second].name, bookingTokens, search::Delimiters());
+
+    map<size_t, vector<pair<size_t, size_t>>> weightPair;
```

Comment:
What's the point to keep this map if all what you need is to check whether it's empty or not.